lib/web_ui/lib/src/engine/compositor/skia_object_cache.dart

```diff
@@ -0,0 +1,238 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of engine;
+
+/// A cache of Skia objects whose memory Flutter manages.
+///
+/// When using Skia, Flutter creates Skia objects which are allocated in
+/// WASM memory and which must be explicitly deleted. In the case of Flutter
+/// mobile, the Skia objects are wrapped by a C++ class which is destroyed
+/// when the associated Dart object is garbage collected.
+///
+/// On the web, we cannot tell when a Dart object is garbage collected, so
+/// we must use other strategies to know when to delete a Skia object. Some
+/// objects, like [ui.Paint], can safely delete their associated Skia object
+/// because they can always recreate the Skia object from data stored in the
+/// Dart object. Other objects, like [ui.Picture], can be serialized to a
+/// JS-managed data structure when they are deleted so that when the associated
+/// object is garbage collected, so is the serialized data.
+class SkiaObjectCache {
+  final int maximumSize;
+
+  final DoubleLinkedQueue<SkiaObject> _itemQueue;
+  final Map<SkiaObject, DoubleLinkedQueueEntry<SkiaObject>> _itemMap;
+
+  SkiaObjectCache(this.maximumSize)
+      : _itemQueue = DoubleLinkedQueue<SkiaObject>(),
+        _itemMap = <SkiaObject, DoubleLinkedQueueEntry<SkiaObject>>{};
+
+  int get length => _itemQueue.length;
+
+  bool contains(SkiaObject object) {
+    return _itemQueue.contains(object);
+  }
+
+  /// Adds [object] to the cache.
+  ///
+  /// If adding [object] causes the total size of the cache to exceed
+  /// [maximumSize], then the least recently used half of the cache
+  /// will be deleted.
+  void add(SkiaObject object) {
+    _itemQueue.addFirst(object);
+    DoubleLinkedQueueEntry<SkiaObject> item = _itemQueue.firstEntry()!;
+    _itemMap[object] = item;
+
+    if (_itemQueue.length > maximumSize) {
+      SkiaObjects.markCacheForResize(this);
+    }
+  }
+
+  /// Records that [object] was used in the most recent frame.
+  void markUsed(SkiaObject object) {
+    DoubleLinkedQueueEntry<SkiaObject> item = _itemMap[object]!;
+    item.remove();
+    _itemQueue.addFirst(object);
+    DoubleLinkedQueueEntry<SkiaObject> newItem = _itemQueue.firstEntry()!;
+    _itemMap[object] = newItem;
```

Comment:
nit: the usage of the local variables is trivial. Inlining the expression might be more readable. In fact, the first expression can even be `_itemMap.remove(object);`, I think.

Reply:
Done.